pkg/entity-store/postgres.go

```diff
@@ -0,0 +1,268 @@
+		return "", err
+	}
+
+	sql := `INSERT INTO entity (key, value, status) VALUES (:key, :value, :status)`
+
+	_, err = p.db.NamedExec(sql, row)
+	if err != nil {
+		return "", errors.Wrap(err, "error adding entity into db")
+	}
+	return id, nil
+}
+
+// Update updates existing entities to the store
+func (p *postgres) Update(lastRevision uint64, entity Entity) (revision int64, err error) {
+
+	entity.setModifiedTime(time.Now())
+	entity.setRevision(lastRevision + 1)
+	sql := `
+	UPDATE entity
+	SET value = $1 ,
+		status = $2
+	WHERE
+		key = $3 AND
+		value::json->>'revision' = $4
+	`
+	row, err := entityToDbEntity(entity)
+	if err != nil {
+		return 0, err
+	}
+	result, err := p.db.Exec(sql, row.Value, row.Status, row.Key, lastRevision)
+	if err != nil {
+		return 0, errors.Wrap(err, "error updating entity into db")
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return 0, errors.Wrap(err, "error updating entity into db")
+	}
+	if rowsAffected != 1 {
+		return 0, errors.Errorf("no such entity or there's intermidate update")
+	}
+	return 0, nil
+}
+
+// Get gets a single entity by key from the store
+func (p *postgres) Get(organizationID string, name string, entity Entity) error {
+
+	key := buildKey(organizationID, getDataType(entity), name)
+
+	sql := `SELECT * FROM entity WHERE key = $1`
+
+	row := dbEntity{}
+	err := p.db.QueryRowx(sql, key).StructScan(&row)
+	if err != nil {
+		return errors.Wrap(err, "error getting entity from db")
+	}
+
+	err = row.Value.Unmarshal(entity)
+	if err != nil {
+		return errors.Wrap(err, "error unmarshalling json from db")
+	}
+	return nil
+}
+
+// List fetches a list of entities of a single data type satisfying the filter.
+// entities is a placeholder for results and must be a pointer to an empty slice of the desired entity type.
+func (p *postgres) List(organizationID string, filter Filter, entities interface{}) error {
+
+	rv := reflect.ValueOf(entities)
+	if entities == nil || rv.Kind() != reflect.Ptr || rv.Elem().Kind() != reflect.Slice {
+		return errors.New("need a non-nil entity slice pointer")
+	}
+
+	elemType := rv.Elem().Type().Elem()
+	if !elemType.Implements(reflect.TypeOf((*Entity)(nil)).Elem()) {
+		return errors.New("non-entity element type: maybe use pointers")
+	}
+
+	key := buildKey(organizationID, dataType(elemType.Elem().Name()))
+
+	sql := `SELECT * FROM entity WHERE key LIKE $1`
+	rows, err := p.db.Queryx(sql, fmt.Sprintf("%s%%", key))
+	if err != nil {
+		return errors.Wrap(err, "error listing entity from db")
+	}
+
+	slice := reflect.MakeSlice(rv.Elem().Type(), 0, 0)
+	for rows.Next() {
+		row := dbEntity{}
+		err = rows.StructScan(&row)
+		if err != nil {
+			return errors.Wrap(err, "error listing entity from db")
+		}
+
+		obj := reflect.New(elemType.Elem())
+		entity := obj.Interface().(Entity)
+		err = row.Value.Unmarshal(entity)
+		if err != nil {
+			return errors.Wrap(err, "error unmarshalling json from db")
+		}
+		// TODO: put filter into sql
```

Comment:
This will get a little tricky with some fields in JSONB and others as columns, but shouldn't be too bad.

Reply:
yea, but I do believe this is the point why we adapt postgres, (the other reason is persistency)